### DISCUSSION TITLE
Microstrip line Simulation

### DISCUSSION THREAD

**Discussion by manujmg** [NanoComp/meep#2743] [Q&A]
This discussion is intended for people who is trying to simulate transmission lines in RF World (mainly focused on microstrip lines)

**Comment by manujmg** [comment]
I am building a basic microstrip model. This comes from the need of people asking about RF examples(e.g. #2493) as well as MPB not supporting metals. 
To simulate a microstrip line, I have created the geometry defined by the below pictures. A microstrip is formed by a ground plane, a dielectric substrate, and a metallic transmission line (usually lossy copper). In order to proper create a microstrip geometry, the S-
Parameters are required to understand what the microstrip performance is. 

<img width="854" alt="Captura de pantalla 2023-12-16 a las 12 23 35" src="https://github.com/NanoComp/meep/assets/26410461/3687d441-be24-43ab-a18a-7226efbf8b09">

<img width="889" alt="Captura de pantalla 2023-12-16 a las 12 23 53" src="https://github.com/NanoComp/meep/assets/26410461/4955498f-2c5b-4460-b40d-967421d7bd11">


In order to get the S-Parameters, I was thinking about creating a Y-Z 2D gaussian source (component in positive Z) large enough to enclose the TEM mode for the microstrip line.  Based on that, a field monitor would be placed somewhere in the microstrip so the fields can be obtained once the simulation is done. Once you get the fields, you could do a mode/decomposition to get the mode profiles, so later you can use those profiles with the GaussianBeam3DSource to excite the microstrip properly.

The question I am having here is.... how S11 would be computed in a scenario like that?

Some portion of the code is below (fair warning, I am using a custom object oriented code which you won't find in MEEP):


   # Global Variables
    a = 1e-3 #1mm characteristic length
    c = 3e8 #speed of light
    f_real = 11.5e9 #Real frequency we want to measure
    time_step = 0.6 #time step for simulation to store ez_values
    epsilon_0 = 8.854187817e-12 #Epsilon zero. 
    dpml = 10  # thickness of PML
    L = 20  # length of non-PML region
    n = 3.6  # refractive index of surrounding medium
    fcen = f_real * a / c #MEEP Frequency center of source/monitor
    simulation_time = 10/fcen #5 times the wave period
    wvl = round(1/fcen,3) # wavelength 
    resolution = (105/wvl) # 100 pixels/wavelength

   # Microstrip Object
    microstrip = Microstrip()
    microstrip.set_ground_parameters(w_sub,w_sub,th_ground,mp.metal)
    microstrip.set_substrate_parameters(w_sub,w_sub,th_sub,rogers)
    microstrip.set_microstrip_parameters(l_feed,w_feed,th_ground)
    microstrip.create_ground(mp.Vector3(0,0,-th_sub*.5 - th_ground*.5))
    microstrip.create_substrate(mp.Vector3(0,0,0))
    microstrip.create_microstrip(mp.Vector3(0,0,th_sub*.5 + th_ground*.5))
     
   # Source Object
    source1 = Source()
    source_center = [-microstrip.w_sub*0.5, 0, -microstrip.th_sub*.5-microstrip.th_ground + (4*microstrip.th_sub)/2]
    source_size = [0,6*microstrip.w_micro,4*microstrip.th_sub]
    source_cmp = mp.Ez
    source1.freq = fcen
    source1.duration = simulation_time
    source1.create_source(2,source_center,source_size,source_cmp)
    source1.compute_tx_power(cell,resolution) #Compute power emitted from source
     
   # Arrays to store power
    incidentFlux_x = []
    
    sim1.run_simulation(saveEz=False)

    #Power from monitor covering an entire surface as a YZ plane in front of the source
    incidentFlux_x.append(sim1.field_monitor.flux_monitors[f'Total{i}'].power)
     
    #FROM HERE I NEED TO DEVELOP THE EXTRACT FIELDS FROM THE MONITOR  - PERFORM A MODE DECOMPOSITION
    #AND USE THOSE MODES TO CREATE A NEW SOURCE (GAUSSIANBEAM3DSOURCE) TO EXCITE THE MICROSTRIP, PERFORM
    #A NEW MODE DECOMPOSITION AND COMPUTE S11 FROM THERE?


Some of the results I got for the distribution of the electric field during the simulation can be viewed in the following gif plot:


https://github.com/NanoComp/meep/assets/26410461/27badea0-cc7c-4084-ae84-0138d9920409

**Reply by manujmg** [reply]
@smartalecH any thoughts?

**Comment by CosimoMV** [comment]
Hi, as far as I know the correct procedure would be as follows (more or like):

- Use an external mode solver to obtain the E and H profiles for the eigenvalues of the structure (forward and backward propagating modes).
- Use a broadband source to excite the various modes over a wide frequency range. For the stripline a good source should be a volume 1d perpendicular to the ground plane, extending from the plane to the strip.
- Decompose the field on the first plane (port one) on the forward and backward propagation modes as shown in the documentation ( https://meep.readthedocs.io/en/latest/Mode_Decomposition/#theoretical-background ) and determine S11.
- Decompose the fields on the second plane (port two) on the forward propagation modes to obtain S12 with the reference of port one.

Calculating the overlap integrals should not be very difficult, but I think the real problem is finding the modes somehow and importing them into meep.

The "funny" thing is that MPB should be able to work with metal structures (negative epsilon), but the problem is that the python interface doesn't permit this.  https://github.com/NanoComp/mpb/issues/135

**Comment by smartalecH** [comment]
> To simulate a microstrip line, I have created the geometry defined by the below pictures. A microstrip is formed by a ground plane, a dielectric substrate, and a metallic transmission line (usually lossy copper). In order to proper create a microstrip geometry, the S-
Parameters are required to understand what the microstrip performance is.

Wait, are you just trying to understand the performance of the microstrip itself? ie you aren't trying to compute the s-parameters of the strip coupling into some arbitrary structure?

If so, *don't* simulate the structure in something like FDTD. You can compute all the complex scattering parameters directly from the mode profiles.

**Reply by CosimoMV** [reply]
Thank you for your response!

> Wait, are you just trying to understand the performance of the microstrip itself? ie you aren't trying to compute the s-parameters of the strip coupling into some arbitrary structure?

I have two coupled microstrip lines and I want to simulate the s-parameters of the strips coupling with a via.
The problem is that I don't know how to find the mode profiles.

**Reply by manujmg** [reply]
@smartalecH no, I am trying to compute the S-Parameters of the strip coupling to a patch antenna suspended above the strip. I am using other commercial tools for comparison and I was not able to find/do a good way to compute the S-Parameters properly with MEEP.

**Comment by stevengj** [comment]
Note that the mode decomposition in Meep (which uses MPB's eigensolver) doesn't currently support metals.  So you'd have to use some other mode solver, and roll your own mode decomposition from the DFT fields, which is theoretically straightforward but rather tedious to get right.
